Enable Swift FastlaneRunner to be updated automatically 🎉

taquitos opened the pull request:
Currently, when you make changes to fastlane API and we regenerate `Fastlane.swift` there is no way to actually get that file into users' `FastlaneRunner` projects.

Not only this file, but also these files (currently):
`Fastlane.swift`,`DeliverfileProtocol.swift`,`GymfileProtocol.swift`,`MatchfileProtocol.swift`,`PrecheckfileProtocol.swift`,`ScanfileProtocol.swift`,`ScreengrabfileProtocol.swift`,`SnapshotfileProtocol.swift`,`LaneFileProtocol.swift`,`ControlCommand.swift`,`RubyCommand.swift`,`RubyCommandable.swift`,`Runner.swift`,`SocketClient.swift`,`SocketClientDelegateProtocol.swift`,`SocketResponse.swift`,`ArgumentProcessor.swift`,`main.swift`, and `RunnerArgument.swift`

These belong to different groups, and as we make changes/fixes we'll be adding other files, and possibly other groups to the Xcode project.

This PR automates the process of updating users' projects so that new files/groups are added, and file changes are applied as well.

This currently only supports overwriting a set of files that are generated, and will not remove/update any of the users' own files.

- Updated deployment script to generate a required Swift file manifest

Current logic is as follows:
- find all groups in existing project
  - if a group is missing, add it
    - add all files for group into new group, build target, and compile phase
- iterate through existing groups
  - update all files needing updating by checking version values first, then checking actual contents
- iterate through existing groups
  - if a file from the manifest is new, add it to group, build target, and compile phase
- Save and return true if any action was taken to modify any file (project included)

Commit:
Upgrade FastlaneSwiftRunner when needed
- during deployment, generate upgrade manifest
- add readme to autogenerated project

Commit:
Only add files to the manifest that the user won't modify

Commit:
current logic:
- find all groups in existing project
-- if a group is missing, add it
--- add all files for group into new group, build target, and compile phase
- iterate through existing groups
-- update all files needing updating
- iterate through existing groups
-- if a file from the manifest is new, add it to group, build target, and compile phase
- Save and return true if any action was taken to modify any file (project included)

Commit:
Remove dev code that was used during development

KrauseFx commented on fastlane/Fastfile:
```diff
@@ -150,6 +154,33 @@ lane(:generate_swift_api) do |options|
+  swift_folder = File.expand_path("swift")
+  runner_project_path = File.expand_path(File.join(swift_folder, "/FastlaneSwiftRunner/FastlaneSwiftRunner.xcodeproj"))
+  project = Xcodeproj::Project.open(runner_project_path)
+  swift_files = project.files.select do |file|
```
Maybe this piece is easier to read if we have multiple `.select do` blocks? It would take longer to run, but this should be done in less than a second anyway, right?

KrauseFx commented on fastlane/lib/fastlane/swift_lane_manager.rb:
```diff
@@ -54,6 +62,11 @@ def self.cruise_lane(lane, parameters = nil, env = nil)
+
+      if display_upgraded_message
+        UI.message("We updated your FastlaneRunner project during this run to make it compatible with your current version of fastlane.")
+        UI.message("Please make sure to check the changes into source control.")
```
What happens after that? That's during a normal _fastlane_ run, right? Will the user be able to see this, or does it pretty much immediately disappear because the actual build output will be printed out?

taquitos replied:
This happens at the end of the run 🎉 

<img width="799" alt="screen shot 2018-01-10 at 10 25 28 am" src="https://user-images.githubusercontent.com/1304821/34788833-a5c890d6-f5f0-11e7-9127-b82d681b658d.png">

KrauseFx commented:
This looks great. I don't think I fully understand the user flow here. 

User runs `fastlane beta` with outdated Swift support files. What happens? From my understanding, unless this flag is used, we automatically upgrade the files, telling the user to check them into git, but as part of the build output, so probably something they might miss.

**Questions:**
- how does this work on CI? Should we always skip the upgrade when `Helper.ci?` is true? As the user will never check it in from there. Also CI shouldn't modify source code
- Instead of automatically upgrading the files for the user, could we show a message at the very end of the _fastlane_ output, telling them to run `fastlane upgrade_swift_files` or similar?

With _fastlane_ we often try to avoid modifying files on behalf of the user automatically, we usually either tell them what command to run, or how to modify certain files/projects to get them where they want to be

taquitos commented:
@KrauseFx you bring up a great point about modifying the user's projects. Traditionally we don't and I totally agree with that.

Here's why I think it's ok for us to do it in this circumstance:
- `FastlaneSwiftRunner.xcodeproj` is a project we supply
  - Any changes will not impact the user's app
  - This project builds a core piece of communication and is 100% required to be updated sometimes
- We have really strong messaging around what files may be overwritten, and what not to do:
  - https://github.com/fastlane/fastlane/pull/11512/files#diff-d4bddcd1602a1495ce67bd9b9faf06d2 is included in every group
- We have an upgrade manifest file that includes the list of files we're going to update, and it's very targeted.
  - Never update configuration files (`Toolnamefile.swift`)
  - Never delete a file
  - Never delete a group
  - Only update files we've generated
- If the user doesn't update, fastlane will likely fail.

I'll respond to your other comments inline

Commit:
Add header to files people shouldn't modify so we can upgrade em

taquitos commented:
> User runs fastlane beta with outdated Swift support files. What happens? 

It could work, but if defaults were changed, params moved, or updated, it will fail on the ruby side. 
The worst case is that we change our protocol and make it backwards incompatible (like I did in 1.74.0). Then the client-server comms would just fail entirely.

> how does this work on CI? 

I forgot about CI, I agree with you:

> always skip the upgrade when Helper.ci? is true

because these changes shouldn't happen at CI time, it should happen in a developer's workflow. I'm going to update it so that we don't do it during CI.

Also, I think I'll add a confirm step to the upgrade with more messaging and with a default of "yes"

revolter commented:
Worst case scenario would be if we change some default values and it __doesn't__ fail but does incorrect things instead, IMO.

Commit:
Use multiple .reject/select to make more readable

Commit:
Don't attempt upgrade at all if running in CI

Commit:
Add dry_run to upgrader so we can see if we would make project changes
Ask user if they'd like to upgrade before doing it

Commit:
Compare FastlaneRunnerAPIVersion of source and target files
so we don't blindly copy files when not needed
- helps during dry run to see if we need to make changes or not

Commit:
Check version numbers of each file and only write out files that need generating
If a generated api needs to be updated, automatically bump the version number
Add failsafe to upgrader, if version numbers are the same but the file content differs, mark for upgrade

Commit:
Bump FastlaneAPIVersion in files that were missed last update
Update comments to express users shouldn't modify auto-gen files or distributed code

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

taquitos commented on fastlane/swift/Fastlane.swift:
```diff
@@ -460,7 +460,7 @@ func buildApp(workspace: String? = nil,
               clean: Bool = false,
               outputDirectory: String = ".",
-              outputName: String? = nil,
+              outputName: String = "TACOS",
```
ugh. test code

Commit:
Undo test code
